### DISCUSSION TITLE
Unable to Auth with AWS Cognito for

### DISCUSSION THREAD

**Discussion by programmrz** [minio/minio#18624] [Q&A] [answered]
I have an existing S3 / Cognito flow that allows an authed user to use an Authorization bearer token to access an S3 file and a policy on the bucket to allow a user to download/access anything in their bucket as long as the authed jwt.sub is the same as the root folder in the bucket. I'd like to move to minio so I set the same policy type, just ALIAS/private/${jwt.sub} but I'm unable to auth with working bearer tokens. 

I am passing the bearer token as I would to auth any other AWS service, and I get 
`The authorization mechanism you have provided is not supported. Please use AWS4-HMAC-SHA256` in postman.

On Minio's end, I've added AWS Cognito settings but will admit im am unsure what scopes/claims to configure in cognito.

* Version used (`minio --version`): RELEASE.2023-11-20T22-40-07Z
* Server setup and configuration: Ubuntu, running Minio. allows 443 which gets sent to Minio
* Operating System and version (`uname -a`): Ubuntu 22.04

**Comment by harshavardhana** [comment]
Does AWS Congito support full OpenID spec ? And it yes how is it configured? 

Can you provide us some more information?

**Reply by harshavardhana** [reply]
This is mostly a discussion moving it here.

**Reply by programmrz** [reply]
Cognito user pools implements ID, access, and refresh tokens as defined by the OpenID Connect (OIDC) open standard, per site. 

Pretty basic... App Client created with the information i'm plugging into minio OpenID settings page. Spun up a duplicate app under the userpool for testing. 

<img width="1082" alt="image" src="https://github.com/minio/minio/assets/8770307/d66b4943-6557-4913-a501-7d97894c52b4">

**Reply by programmrz** [reply]
is there a way to see what credentials were attempted on a bucket?. Like a jwt.sub was ey123 but since policy is expecting ey456 it was denied?

**Reply by harshavardhana** [reply]
Can you share a sample JWT you get from AWS Cognito?

**Reply by programmrz** [reply]
eyJraWQiOiJSRjM5VGw2S0s5Nk9cL0pyMm41eUhKVGVtZzFFNjh1ZHU0NlRuVnAxMEU4ST0iLCJhbGciOiJSUzI1NiJ9.eyJzdWIiOiI2NGI4NzQ4OC1iMGYxLTcwYjUtMzFlMC01MTc0OWJmODc3M2UiLCJjb2duaXRvOmdyb3VwcyI6WyJVc2VycyJdLCJpc3MiOiJodHRwczpcL1wvY29nbml0by1pZHAudXMtZWFzdC0xLmFtYXpvbmF3cy5jb21cL3VzLWVhc3QtMV8yV3V3ajNvUnEiLCJjbGllbnRfaWQiOiI0bmkxdjJ0aGZ2a2xsNTJjb25lbmM0anU1MiIsIm9yaWdpbl9qdGkiOiI2OWE3YzlkNC0yMTgxLTRkNjYtOTY5YS03ZTQ5ODFiNzNlNmMiLCJldmVudF9pZCI6IjhhYjAwNWVhLTMwMjctNGY2Zi04NDVkLWE5MDE5MzJhMGUxZCIsInRva2VuX3VzZSI6ImFjY2VzcyIsInNjb3BlIjoiYXdzLmNvZ25pdG8uc2lnbmluLnVzZXIuYWRtaW4iLCJhdXRoX3RpbWUiOjE3MDIzMTA3NzIsImV4cCI6MTcwMjM5NzE3MiwiaWF0IjoxNzAyMzEwNzcyLCJqdGkiOiJhNTdhMWM3ZC01ODQyLTQxNzMtOWIzZi0zNDE3NzM3YTZlYTEiLCJ1c2VybmFtZSI6IjY0Yjg3NDg4LWIwZjEtNzBiNS0zMWUwLTUxNzQ5YmY4NzczZSJ9.fMbaR0d3fJ6dXee0axA04AoDSlPkm1_5pYEEYi0YQ4H9Ec7qlXYs5BTaDGUu7VA61Wtevgx5vOhcW8Q1p6DWeFSnWN2rKODPqSKLCArFLxG2Xu2QNdmZz2iNN72v5StTBYbLW91Zc2k5kO_dw3iKyU_0UtYWjRkCkFXUJIlOohxY0FSlXGWwGWaTthmWs08TtZz6wMfvOEYaRBl1ex1UkUJckk-TldwKJkau64KOBagGRpzKgjRBfdILnsLwFoM5t-GYWyh9Kz6mcU-8dKYBmPKhCzM9-u4oLD7xhy3oXkopG_liTGEIJifDcuts-Z-eEmXa-Ra-qVOdaXbKpqyw2g

**Reply by harshavardhana** [reply]
Suppose you don't provide a direct policy to the IDP configuration. In that case, MinIO expects you to have a claim_name that we can use to offer policies to the newly generated credentials.

```json
{
  "kid": "RF39Tl6KK96O/Jr2n5yHJTemg1E68udu46TnVp10E8I=",
  "alg": "RS256"
}.{
  "sub": "64b87488-b0f1-70b5-31e0-51749bf8773e",
  "cognito:groups": [
    "Users"
  ],
  "iss": "https://cognito-idp.us-east-1.amazonaws.com/us-east-1_2Wuwj3oRq",
  "client_id": "4ni1v2thfvkll52conenc4ju52",
  "origin_jti": "69a7c9d4-2181-4d66-969a-7e4981b73e6c",
  "event_id": "8ab005ea-3027-4f6f-845d-a901932a0e1d",
  "token_use": "access",
  "scope": "aws.cognito.signin.user.admin",
  "auth_time": 1702310772,
  "exp": 1702397172,
  "iat": 1702310772,
  "jti": "a57a1c7d-5842-4173-9b3f-3417737a6ea1",
  "username": "64b87488-b0f1-70b5-31e0-51749bf8773e"
}.[Signature]
```

It looks like you need to provide 

```
claim_name="cognito:groups"
```

**Comment by harshavardhana** [comment]
> I am passing the bearer token as I would to auth any other AWS service, and I get
> `The authorization mechanism you have provided is not supported. Please use AWS4-HMAC-SHA256` in postman.

That's different from how you can use this API, you need to use AssumeRoleWithWebIdentity API and then you get STS (rotating credentials) that you can use with MinIO.

**Reply by programmrz** [reply]
understood. I was able to modify my Cognito Token to include the policy claim and was able to send it to AssumeRoleWithWebIdentity. I was given a id, secret and session token, but that session token still gives the same AWS4 error. The hope is to auth an image with just the url and a bearer header

This is the current session token im passing in Authorization: Bearer <token>

eyJhbGciOiJIUzUxMiIsInR5cCI6IkpXVCJ9.eyJhY2Nlc3NLZXkiOiJOWlBWMFZGWldPNE5LR0FaRFJZSiIsImF1ZCI6IjRuaTF2MnRoZnZrbGw1MmNvbmVuYzRqdTUyIiwiYXV0aF90aW1lIjoxNzAyMzUzMzc2LCJjb2duaXRvOmdyb3VwcyI6WyJVc2VycyJdLCJjb2duaXRvOnVzZXJuYW1lIjoiNjRiODc0ODgtYjBmMS03MGI1LTMxZTAtNTE3NDliZjg3NzNlIiwiZW1haWwiOiJwZXRhQGh5cGVybGluay5tZSIsImVtYWlsX3ZlcmlmaWVkIjp0cnVlLCJldmVudF9pZCI6IjAxZmM5MmU3LWUzMGEtNDdhNC1hYWMzLTVjMzA0ZGQ2Zjg1MCIsImV4cCI6MTcwMjQzOTc3NSwiZm9sZGVyIjoieW91cnMiLCJpYXQiOjE3MDIzNTMzNzYsImlzcyI6Imh0dHBzOi8vY29nbml0by1pZHAudXMtZWFzdC0xLmFtYXpvbmF3cy5jb20vdXMtZWFzdC0xXzJXdXdqM29ScSIsImp0aSI6IjhkMmEwNDRhLWNjMzgtNDJmZS05ZDM4LTYxZjUwNjkzNjljOSIsIm9yaWdpbl9qdGkiOiIzNGMxYTU2OC1kZjg0LTQyOGYtYTUxOC1lODQ4Y2FlZTdjMTQiLCJwb2xpY3kiOiJ1c2VyIiwic3ViIjoiNjRiODc0ODgtYjBmMS03MGI1LTMxZTAtNTE3NDliZjg3NzNlIiwidG9rZW5fdXNlIjoiaWQifQ.RBUYjBSLsLJEvLuEUTdK3lz7RgJ3YCVp7veUkd3-U0C9u9E39bQXrrGSoswmjGRqZArEKlGnhDNf_IMLOBKnKw

**Reply by harshavardhana** [reply]
Once you have those you can't just send bearer token to MinIO. You need to use an S3 client to talk to MinIO that implements signature v4.

**Reply by harshavardhana** [reply, marked as answer]
> AssumeRoleWithWebIdentity. I was given a id, secret and session token, but that session token still gives the same AWS4 error. The hope is to auth an image with just the url and a bearer header

I think you don't understand how to make API calls via S3 API. You should read about how all three play a role in authentication.

- access key
- secret key
- session token

**Reply by programmrz** [reply]
ahhh, I see. I'm considering token as a bearer when it's not. 

I think i'm heading in the right direction, I will expand on [this](https://stackoverflow.com/questions/72047783/how-do-i-download-files-from-a-minio-s3-bucket-using-curl) in order to meet my needs. Thank you for the help. I tend to start building before getting an understanding of the instructions lol

**Reply by programmrz** [reply]
in case someone is searching, it's possible to follow the above steps to create a rotating STS credentials from AWS Cognito without an app secret (you'll need to use env file, just dont include the openid secret). Once you have your key, token and id, you can sign your url request to receive headers. You can then pass these headers to your <Image> component and now you can access/cache content dynamically 
<img width="742" alt="image" src="https://github.com/minio/minio/assets/8770307/0428b36e-89b4-47da-8ee0-9a700451eeae">
